### DISCUSSION TITLE
`The NVIDIA driver's CUDA version is 12.2 which is older than the ptxas CUDA version (12.3.107). Because the driver is older than the ptxas version, XLA is disabling parallel compilation, which may slow down compilation. You should update your NVIDIA driver or use the NVIDIA-provided CUDA forward compatibility packages.`

### DISCUSSION THREAD

**Discussion by VachanVY** [jax-ml/jax#20156] [Q&A]
Hi,
getting this message when importing jax
> 2024-03-10 21:46:26.964590: W external/xla/xla/service/gpu/nvptx_compiler.cc:698] The NVIDIA driver's CUDA version is 12.2 which is older than the ptxas CUDA version (12.3.107). Because the driver is older than the ptxas version, XLA is disabling parallel compilation, which may slow down compilation. You should update your NVIDIA driver or use the NVIDIA-provided CUDA forward compatibility packages.

what should I do to get rid of this message?

**Comment by jakevdp** [comment]
Hi - thanks for the question! The error lays out the issue pretty clearly I think: you have CUDA 12.2 installed, and JAX requires CUDA 12.3. To fix this, you should upgrade your CUDA libraries. Can you say more about how you installed JAX?

**Reply by VachanVY** [reply]
I installed jax using the command 
`pip install --upgrade "jax[cuda12_pip]" -f https://storage.googleapis.com/jax-releases/jax_cuda_releases.html`. 
And can you please guide me on how can I upgrade cuda? Can I follow [this answer?](https://stackoverflow.com/questions/50213021/best-practice-for-upgrading-cuda-and-cudnn-for-tensorflow#:~:text=First%2C%20I%20uninstalled-,cuda,-and%20all%20its)

So is that the reason why I was getting out of memory error when I jit the `train_step` function and why it's soo slow without jitting it?
Thanks.

**Reply by jakevdp** [reply]
That command will install the correct CUDA versions. If you're still getting an error about incorrect CUDA versions, it means there are other CUDA libraries installed in your system. I would suggest double-checking that your `LD_LIBRARY_PATH` is pointing to the set of CUDA libraries installed by your `pip` command.

**Reply by VachanVY** [reply]
> I would suggest double-checking that your LD_LIBRARY_PATH is pointing to the set of CUDA libraries installed by your pip command.

Sorry, I didn't understand. Can you please elaborate?

(Also, how can I revert back to a version of JAX that uses a specific version of CUDA?)
(And I checked Nvidia's website but didn't find any webpage for upgrading cuda)

**Reply by jakevdp** [reply]
> Sorry, I didn't understand. Can you please elaborate?

When your process loads CUDA, it needs to find the CUDA libraries that you've installed on your system. The way you specify where your system will look for these libraries is via a system variable called `LD_LIBRARY_PATH`. Often errors like the one you have, which complain of mismatch between versions of CUDA components, arise in situations where there are multiple copies of CUDA libraries installed on the system, and the `LD_LIBRARY_PATH` variable is set in such a way that the system loads conflicting versions.

Does that help?

**Reply by VachanVY** [reply]
Ok, so how should I proceed with checking the `LD_LIBRARY_PATH`? Any commands or something?

**Reply by jakevdp** [reply]
You can open a system prompt and type
```
echo $LD_LIBRARY_PATH
```
There's a bit more information here: https://jax.readthedocs.io/en/latest/installation.html#pip-installation-gpu-cuda-installed-via-pip-easier

**Reply by VachanVY** [reply]
after typing in that command on the terminal I don't get anything, just a newline.

Does the below help?
```
python3.11 -m pip list | grep nvidia
nvidia-cublas-cu12           12.3.4.1
nvidia-cuda-cupti-cu12       12.3.101
nvidia-cuda-nvcc-cu12        12.3.107
nvidia-cuda-nvrtc-cu12       12.1.105
nvidia-cuda-runtime-cu12     12.3.101
nvidia-cudnn-cu12            8.9.7.29
nvidia-cufft-cu12            11.0.12.1
nvidia-curand-cu12           10.3.2.106
nvidia-cusolver-cu12         11.5.4.101
nvidia-cusparse-cu12         12.2.0.103
nvidia-nccl-cu12             2.19.3
nvidia-nvjitlink-cu12        12.3.101
nvidia-nvtx-cu12             12.1.105
```

**Reply by jakevdp** [reply]
OK, then `LD_LIBRARY_PATH` is not defined, and you should be loading cuda from the pip installed packages. It looks like some of those packages are older than they should be; I'd suggest running the pip install command again:
```
pip install -U jax[cuda12_pip] -f https://storage.googleapis.com/jax-releases/jax_cuda_releases.html
```

**Reply by VachanVY** [reply]
The issue is not solved, I still get that msg and when I type `nvidia-smi` I don't know why it still shows the old Cuda version 12.2 (as in the question) instead of the one that Jax installs.

So I just removed the old Nvidia driver and downloaded the new one from their website but in the Additional drivers (in Ubuntu) it still shows me the old driver version

Is there a way to just get the jax install for Cuda 12.2 (the old version which is showing for me)

**Reply by jakevdp** [reply]
jax v0.4.24 is compatible with cuda 12.2. But I suspect the root of the issue is that you have multiple incompatible sets of CUDA installations on your machine (one set that you installed from the nvidia website, one set that you installed via `pip`). I would suggest removing one or the other.

**Reply by jakevdp** [reply]
Note also that this warning is not necessarily a blocker; see https://github.com/google/jax/issues/18027#issuecomment-1758108802

**Reply by realjoenguyen** [reply]
@jakevdp I got the same issue 
Here is my `LD_LIBRARY_PATH` 

`/usr/local/apps/cuda/12.2/cudnn/lib:/usr/local/apps/cuda/12.2/targets/x86_64-linux/lib:/usr/local/apps/cuda/12.2/lib64:/nfs/hpc/share/nguyejoe/anaconda3/envs/dynalang_py10/lib/python3.8/site-packages/tensorrt_libs:/usr/local/apps/cudnn/8.9_cuda12/lib::
`

Here is my jax and jaxlib versions:

```
jax==0.4.28
jax-cuda12-pjrt==0.4.23
jax-cuda12-plugin==0.4.23
jaxlib==0.4.28+cuda12.cudnn89
```

**Reply by kvablack** [reply]
Hi @jakevdp, I'd like to revisit this. I realize that this is not a blocker, but it would be very nice to get faster compile times, as I find they can explode shockingly fast.

As #18027 clarifies, this is not about the installed CUDA version(s), but the *NVIDIA driver version*, which is why @realjoenguyen has not been able to solve the problem by changing things about their CUDA installation. It seems like the core issue here is that jaxlib aggressively bumps CUDA and only supports the most recent version (now 12.5), whereas most people are not nearly as aggressive about upgrading their NVIDIA driver. In many cases, such as mine, it's not even possible --- the NVIDIA driver version is pinned by the cloud service provider (and for good reason). Currently, NVIDIA driver 535 (supporting CUDA 12.2) seems to be pretty standard, meaning you would have to go back to JAX 0.4.23 to get parallel compilation.

I don't know much about CUDA, so I have no idea how easy it is to support older CUDA versions. I realize this might be an unreasonable ask. However, I wanted to flag this as a very real problem that a lot of JAX users probably have.

**Reply by kvablack** [reply]
Oh, I should also add that this is solvable by installing an NVIDIA forward compatibility package (as the error message suggests). However, I tried this and got some random segfaults many hours into training, which scared me into uninstalling it pretty quick.

**Reply by hawkinsp** [reply]
We actually have plans to support parallel compilation without the driver. It's just blocked on us migrating across the board to CUDA 12.4 to work around an NVIDIA bug. So hopefully you'll stop seeing this warning in the next few weeks.

**Reply by hawkinsp** [reply]
Incidentally, usually on cloud you *can* choose a NVIDIA driver inside your VM freely. At least, that's the way it works on GCP, I can't speak to other vendors.

**Reply by kvablack** [reply]
Oh that's great news, thanks! Looking forward to it :)

**Reply by quanvuong** [reply]
friendly ping to see if we have parallel compilation without the driver yet. Thank you!

**Comment by xujiboy** [comment]
I am getting similar warning on a Amazon Linux 2 machine:
```
2024-08-07 01:22:52.431656: W external/xla/xla/service/gpu/nvptx_compiler.cc:760] The NVIDIA driver's CUDA version is 12.4 which is older than the ptxas CUDA version (12.6.20). Because the driver is older than the ptxas version, XLA is disabling parallel compilation, which may slow down compilation. You should update your NVIDIA driver or use the NVIDIA-provided CUDA forward compatibility packages.
```
But when I check `ptxas --version` I got the following:
```
ptxas: NVIDIA (R) Ptx optimizing assembler
Copyright (c) 2005-2024 NVIDIA Corporation
Built on Thu_Mar_28_02:14:54_PDT_2024
Cuda compilation tools, release 12.4, V12.4.131
Build cuda_12.4.r12.4/compiler.34097967_0
```
which indicates that ptxas is also on CUDA 12.4. This conflicts with the warning "The NVIDIA driver's CUDA version is 12.4 which is older than the ptxas CUDA version (12.6.20)". Am I missing something?

**Reply by heleidsn** [reply]
same issue
The NVIDIA driver's CUDA version is 12.4 which is older than the ptxas CUDA version (12.5.40).

**Reply by MalavikaVasist** [reply]
same issue. 
 The NVIDIA driver's CUDA version is 12.4 which is older than the ptxas CUDA version (12.6.68)
But `ptxas --version` is, 
```
ptxas: NVIDIA (R) Ptx optimizing assembler
Copyright (c) 2005-2024 NVIDIA Corporation
Built on Tue_Feb_27_16:15:50_PST_2024
Cuda compilation tools, release 12.4, V12.4.99
Build cuda_12.4.r12.4/compiler.33961263_0
```

**Reply by claireyyang** [reply]
I ran into pretty much the same warning (NVIDIA driver's CUDA version is 12.4 which is older than the ptxas CUDA version (12.6.85) and was able to resolve it by running `pip install --upgrade "jax[cuda12]"`, since I installed jax via pip and am not tied to a particular jax version. 

It seems like all it did was upgrade some of the pip dependencies, since it printed out `Successfully installed jax-0.4.35 jax-cuda12-pjrt-0.4.35 jax-cuda12-plugin-0.4.35 jaxlib-0.4.34 nvidia-cudnn-cu12-9.5.1.17`. The warning isn't showing up anymore at least anymore, so maybe this may help someone else who also runs across this issue.
